Lesson 4

kartsevnik opened the pull request.

kartsevnik committed:
rewrite style body, change logic input search, add dropdown categories

kartsevnik committed:
add routing

kartsevnik committed:
add hover, change logo

kartsevnik committed:
add detail movie menu

inga-lekhman commented on src/app/app-routing.module.ts:
```diff
@@ -1,10 +1,34 @@
+  { path: 'popular', canActivate: [MovieGuard], component: PopularComponent },
+  { path: 'top-rate', canActivate: [MovieGuard], component: TopRateComponent },
+  { path: 'upcoming', canActivate: [MovieGuard], component: UpcomingComponent },
+  // { path: 'movie/:id', canActivate: [MovieGuard], component: MovieListViewComponent, resolve: {data: MovieResolver} },
```
Це би теж треба було б реалізувати

inga-lekhman commented on src/app/app.component.html:
```diff
@@ -1,5 +1,7 @@
-</div>
-<!-- <router-outlet /> -->
\ No newline at end of file
+    <!-- <div class="wrapper__sidebar"> <app-sidebar></app-sidebar></div> -->
```
Старайтесь форматувати файл при збереженні (забирати лишні пробіли, пусті рядки, ставити крапку з комою в кінці і т.д.). Це можна налаштувати в IDE.
Тоді ваш файл буде чистим, читабельним і легко підтримуваним.

Це стосується всіхз файлів.
В такому разі ви будете просто насолоджуватися вашим кодом, так само, як і ваші колеги )

inga-lekhman commented on src/app/components/header/header.component.scss:
```diff
@@ -62,4 +68,183 @@ li {
+    animation: fadeOutLeft 1s ease-in-out forwards;
+}
+
+@keyframes fadeInLeft {
```
Такі функції краще виносьте в глобальний файл і просто тут його підключайте. В такому разі ви уникнете дублювання коду. І код буде більш організований

inga-lekhman commented on src/app/components/header/header.component.scss:
```diff
@@ -62,4 +68,183 @@ li {
+}
+
+
+.ramka-3 {
```
Також користуйтеся вкладеністю. 

В загальному, робіть ваші файли настільки маленькими, наскільки це можливо: можливо щось винести або розділити - робіть це. 
Це хороша практика і це робить ваш проект максимально підтримуваним!

inga-lekhman commented on src/app/services/data.service.ts:
```diff
@@ -0,0 +1,43 @@
+  ]
+
+  movies: movie[] = [
+    new movie(1, "The Shawshank Redemption", 1994, 142, "assets/img-movies/The Shawshank Redemption.jpg", false, false, 9.3, "Two imprisoned men bond over a number of years, finding solace and eventual redemption through acts of common decency.", "HD", ["Drama", "Crime"], "/assets/img-movies/The Shawshank Redemption.jpg",),
```
Такі речі краще виносити в окремий файл. Це покращить читабельність цього файлу, дозволить використати ці фейкові дані десь-інде.

inga-lekhman commented on src/app/services/data-handler.service.ts:
```diff
@@ -0,0 +1,127 @@
+import { Injectable } from '@angular/core';
```
Я бачу, ви використали тут щось з інтернету. 
Насправді ви могли просто зробити так, як я показувала. 

Логіка в цьому сервісі насправді дуже складна і не підтримувана в подальшому:
1) не потрібно завантажувати дані при створенні цього сервісу - дані потрібно завантажувати тоді, коли користувач переходить на котрусь сторінку, а не коли створюється сервіс.
2) незрозумло, навіщо вам оновлювати вибрані фільми - я не бачу, який сенс цього у вашому додатку. Можливо ви щось не доробили
3) Такі методи як **forceUpdateCategory** не рекомендується робити.


Тому я попрошу вас видалити цей сервіс, а всю логіку перенести в data.service.
Цей сервіс має:
1) зберігати **favorites** and **watchList** фільми в себе, і мати методи для отримання їх у відповідних компонентах
2) мати методи для додавання та видалення цих фільмів з відповідних категорій (такі як ви маєте тут)
3) повертати фільми по категоріям та фільм по айді

Все максимально просто. Спробуйте зробіть, будь ласка! 
І тоді ви самі будете максимально задоволені. Бо тут таке враження, ніби ви скопіювали десь це і використали не дуже по призначенню. Робіть як я показувала на уроці.

inga-lekhman commented on src/app/components/movie-card/movie-card.component.scss:
```diff
@@ -133,9 +146,116 @@
 .selected-toWatch {
-    background-color: hsla(0, 100%, 50%, 1.0);
+    background-color: var(--maincolor);
+}
```
Користуйтеся вкладеністю у файлах стилів, це дуже зручно, а також підвищує читабельність та редагування стилів.
також виносьте стилі, які відносяться до певного компонента, тільки в стилі цього компонента.
В **Angular** стилі ізольовані для кожного компонента, тому це важливо.

inga-lekhman commented on src/app/components/movie-list/movie-list.component.html:
```diff
@@ -1,32 +1,37 @@
-<ng-container *ngIf="selectedCategory == 'To Watch'">
-    <div class="movie-list-cards ToWatch">
-        <h3>To watch</h3>
+<ng-container
```
Просто так не використовуйте ng-container.
Він зазвичай потрібен, коли на одному елементі вам треба використати дві структурні директиви (*ngIf, *ngSwitch для прикладу)

inga-lekhman commented on src/app/components/movie-list/movie-list.component.ts:
```diff
@@ -1,71 +1,59 @@
 
+  constructor(private dataHandlerService: DataHandlerService) {}
+  ngOnInit(): void {
+    this.movieForHomeBlock = this.selectedMovies[2];
```
Не зрозуміло, чому тут 2-ий елемент юзається. Також пропоную вам видалити цю логіку.

inga-lekhman commented on src/app/components/movie-list/movie-list.component.ts:
```diff
@@ -1,71 +1,59 @@
-  favoriteMovies: movie[] = []
-  toWatchMovies: movie[] = []
+  // получаем текущую категорию из сервиса
+  get selectedCategory() {
```
У вас багато логіки, яку я не просила робите. Я би вам радила повидаляти все непотрібне, оскаільки буде важчу потім робити наступні домашні завдання.

inga-lekhman commented on src/app/components/movie-list/movie-list.component.ts:
```diff
@@ -1,71 +1,59 @@
+  }
+
+  shouldDisplayMovieList(): boolean {
+    const categoriesToShow = ['Home', 'All Movies', 'Favorites', 'To Watch'];
```
Home сторінка у вас нічого не показує, тому не розумію, навіщо вона. Ви могли на ній показувати **All films** наприклад.

inga-lekhman commented on src/app/components/header/header.component.ts:
```diff
@@ -1,26 +1,126 @@
+    this.dataHandlerService.changeCategory('Home');
+
+    // Add document click listener
+    document.addEventListener('click', this.onDocumentClick.bind(this));
```
Таке неприпустиме тут ) це чистий яваскрипт. Вам потрібно робити все в **Angular** стилі.

inga-lekhman commented on src/app/components/header/header.component.ts:
```diff
@@ -1,26 +1,126 @@
+  // при создании компонента загрузка списка категорий и выбор категории по умолчанию для загрузки
+  ngOnInit(): void {
+    this.categoryList = this.dataService.categoryList;
+    this.dataHandlerService.changeCategory('Home');
```
Вам зміну сторінок потрібно робити через **routerLink**. Ви тут не зовсім правильно використали роутінг.

inga-lekhman commented on src/app/components/header/header.component.html:
```diff
@@ -1,28 +1,39 @@
+                    placeholder="Select a Category" class="dropdown-categories__item"
+                    (onChange)="changeCategory($event.value.code)" />
+            </div>
+            <button class="header-icon-block__icon favorites-icons" (click)="changeCategory('Favorites')"><img
```
Замість **click - routerLink**, на який потрібно перейти згідно app-routing.module конфігурації.

inga-lekhman reviewed (commented):
В загальному, якщо це 4-5 домашка, то тут мав би бути роутінг + бібліотека. 
У з бібліотеки мало що використовується, здається тільки кнопки. 
Ви би могли використати **p-card** точно для карточок фільму - тоді б вам пришлось писати набагато менше **HTML** + **SCSS**. В **PrimeNg** вже все готово для вас, саме в цьому і перевага уже готових бібліотек.

на рахунок роутінгу - ви не правильно його застосували - перехід мід сторінками має виконуватись через **routerLink**, або через **ActivaterRoute** клас.

Ви додали багато логіки, однак це дуже ускладнює даний проект. 
Якщо ви зробите все в **Angular** стилі, ви побачите, наскільки зменшитсья ваш код і наскільки він стане зрозумілішим.

Спробуйте робите все, як було на уроці!

kartsevnik committed:
lesson 6-7

- install /@angular/cdk
- change logic routing
- change header (delete dropdown, add sidebar)
- add home component
- change method get movie in detail view by id
- refactored and modificated all services
- add mock data
